File: boson-scala/src/main/scala/io/zink/boson/impl/BosonValidate.scala
package io.zink.boson.impl

import java.nio.ByteBuffer

import io.zink.boson.Boson
import io.zink.boson.bson.bsonImpl.BosonImpl
import io.zink.boson.bson.bsonPath.{DSLParser, Interpreter}
import scala.util.{Failure, Success}
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future


class BosonValidate[T](expression: String, validateFunction: T => Unit) extends Boson{

//  private def callParse(boson: BosonImpl, expression: String): Unit = {
//    val parser = new DSLParser(expression)
//    try {
//      parser.Parse() match {
//        case Success(result) =>
//          new Interpreter[T](boson, result, fExt = Option(validateFunction)).run()
//        case Failure(exc) => throw exc
//      }
//    } catch {
//      case e: RuntimeException => throw new Exception(e.getMessage)
//    }
//  }

  override def go(bsonByteEncoding: Array[Byte]): Future[Array[Byte]] = {
    Future{
//      val boson:BosonImpl = new BosonImpl(byteArray = Option(bsonByteEncoding))
//        callParse(boson,expression)
        bsonByteEncoding
    }
  }


//  override def fuse(boson: Boson) = ???
  //new BosonFuse(this,boson)
  override def go(bsonByteEncoding: String): Future[String] = ???
}

File: boson-java/src/main/java/io/zink/boson/impl/BosonValidate.java
package io.zink.boson.impl;

import io.zink.boson.bson.bsonImpl.BosonImpl;
import io.zink.boson.bson.bsonPath.*;
import io.zink.boson.Boson;
import scala.Function1;
import scala.Option;
import scala.runtime.BoxedUnit;
import scala.util.Try;
import scala.util.parsing.combinator.Parsers;

import java.nio.ByteBuffer;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;
import java.util.function.Function;

public class BosonValidate<T> implements Boson {

    private String expression;
    private Consumer<T> validateFunction;
    private Function1<T, BoxedUnit> anon;

    public BosonValidate(String expression, Consumer<T> validateFunction) {
        this.expression = expression;
        this.validateFunction = validateFunction;
        this.anon = new Function1<T, BoxedUnit>() {
            @Override
            public BoxedUnit apply(T v1) {
                validateFunction.accept(v1);
                return BoxedUnit.UNIT;
            }
        };
    }

//    private void callParse(BosonImpl boson, String expression){
//        DSLParser parser = new DSLParser(expression);
//        try{
//            Try<ProgStatement> pr = parser.Parse();
//            if(pr.isSuccess()){
//                Interpreter interpreter = new Interpreter<>(boson, pr.get(), Option.empty(), Option.apply(anon));
//                interpreter.run();
//            }else{
//                throw new RuntimeException("Failure/Error parsing!");
//            }
//        }catch (RuntimeException e){
//            throw new RuntimeException(e.getMessage());
//        }
//    }

    @Override
    public CompletableFuture<byte[]> go(byte[] bsonByteEncoding) {
        return CompletableFuture.supplyAsync(() -> {
//            Option<byte[]> opt = Option.apply(bsonByteEncoding);
//            Option e = Option.empty();
//            BosonImpl boson = new BosonImpl(opt, e,e);
//            try {
//                callParse(boson, expression);
//                return bsonByteEncoding;
//            } catch (Exception ex) {
//                validateFunction.accept(null);
//                return null;
//            }
            return bsonByteEncoding;
        });
    }

    @Override
    public CompletableFuture<String> go(String bsonByteBufferEncoding) {
        return CompletableFuture.supplyAsync(() -> {
//            Option<ByteBuffer> opt = Option.apply(bsonByteBufferEncoding);
//            Option e = Option.empty();
//            BosonImpl boson = new BosonImpl(e, opt,e);
//            try {
//                callParse(boson, expression);
//                return bsonByteBufferEncoding;
//            } catch (Exception ex) {
//                validateFunction.accept(null);
//                return null;
//            }
            return bsonByteBufferEncoding;
        });
    }

//    @Override
//    public Boson fuse(Boson boson) {
//        return null;
//    }
}
